docs/design/datacontracts/data_descriptor.md

```diff
@@ -0,0 +1,310 @@
+    uint16_t Size;
+};
+
+struct FieldSpec
+{
+    uint32_t Name;
+    uint16_t FieldOffset;
+    uint16_t Reserved;
+};
+
+struct GlobalSpec
+{
+    uint32_t Name;
+    uint32_t Reserved;
+    uint64_t Value;
+};
+```
+
+Rationale: the blob should be producable by including a specially formatted C header file multiple
+times with redefinitions of some macro names appearing in its content.  Additionally the blob avoids
+pointers so that it may be read off from the on-disk representation of an object file.
+
+The blob begins with a directory that gives the relative offsets of the `Types`, `FieldPool`,
+`GlobalValues` and `Names` fields of the blob.  The number of elements of each of the arrays is
+next. This is followed by the sizes of the `TypeSpec`, `FieldSpec` and `GlobalSpec` structs.
+
+Rationale: If a `BinaryBlobDataDescriptor` is created via C macros, we want to embed the `offsetof`
+and `sizeof` of the components of the blob into the blob itself without having to account for any
+padding that the C compiler may introduce to enfore alignment.
+
+The baseline is specified as an offset into the names pool.
+
+The types are given as an array of `TypeSpec` elements.  Each one contains an offset into the
+`NamesPool` giving the name of the type, An offset into the fields pool indicating the first
+specified field of the type, and the size of the type in bytes or 0 if it is indeterminate.
+
+The fields pool is given as a sequence of `FieldSpec` elements.  The fields for each type are given
+in a contiguious subsequence and are terminated by a marker `FieldSpec` with a `Name` offset of 0.
+(Thus if a type has an empty sequence of fields it just points to a marker field spec directly.)
+For each field there is a name that gives an offset in the name pool and an offset indicating the
+field's offset.  The field type is not given.
+
+Rationale: it is expected that the types of the fields were provided by a "baseline" data descriptor.
+
+The globals are gives as a sequence of `GlobalSpec` elements.  Each global has a name and a value.
+The types of the globals are not given.
+
+Rationale: it is expected that the types of the global values were provided by a "baseline" data descriptor.
+
+The `NamesPool` is a single sequence of utf-8 bytes comprising the concatenation of all the type
+field and global names including a terminating nul byte for each name.  The same name may occur
+multiple times.  The names will be referenced by multiple type or multiple fields. (That is, a
+clever blob emitter may pool strings).  The first name in the name pool is the empty string (with
+its nul byte).
+
+Rationale: we want to reserve the offset 0 as a marker.
+
+Names are referenced by giving their offset from the beginning of the `NamesPool`.  Each name
+extends until the first nul byte encountered at or past the beginning of the name.
+
+
+## Example
+
+And example C header describing some data types is given in [sample.data.h](./sample.data.h). And
```

Comment:
In the resulting .o file it will look the same. it can be produced by other means - it doesn't need to be produced by C marcros.  (In fact even for CoreCLR it doesn't need to be produced by macros.  It can be created with C++ `constexpr` and templates).  The preprocessor example is just meant to demonstrate that this can be constructed using C for the most restricted runtimes.

Reply:
> it doesn't need to be produced by C marcros

I wasn't worried that it would be constrained to macros :) I just have no understanding of what alternative method we would use to generate it instead and I'm searching for info that helps me fill in that gap.

Reply:
Yes, I would expect a variant of `cdac-build-tool` to be part of the ILCompiler in the fullness of time.